### ISSUE TITLE
Resolution strategy overrides effects of alignments when used `belongsTo`

### ISSUE THREAD

**Issue by chali** [nebula-plugins/gradle-nebula-integration#11]
Example of difference when resolution strategy is applied. The whole project is here https://github.com/nebula-plugins/gradle-nebula-integration/tree/master/alignmentRulesAndResolutionStrategy

With nebula implementation, a group is aligned in the end a rule effect is overridden.

Call: `./gradlew dependencyInsight --configuration compile --dependency test.nebula`

```
test.nebula:a:1.0.0
   variant "default" [
      org.gradle.status = release (not requested)
   ]
   Selection reasons:
      - Was requested
      - Selected by rule : aligned to 1.0.0 by rules
                with reasons: nebula.resolution-rules uses: rules.json

test.nebula:a:1.0.0
\--- compile

test.nebula:b:1.0.0
   variant "default" [
      org.gradle.status = release (not requested)
   ]
   Selection reasons:
      - Was requested
      - Selected by rule : aligned to 1.0.0 by rules
                with reasons: nebula.resolution-rules uses: rules.json

test.nebula:b:0.15.0 -> 1.0.0
\--- compile

test.nebula:c:1.0.0
   variant "default" [
      org.gradle.status = release (not requested)
   ]

test.nebula:c:1.0.0
\--- compile
```

With gradle core based implementation module `a` is brought down by a rule and it causes misaligned group.

Call: `./gradlew dependencyInsight --configuration compile --dependency test.nebula -Dnebula.features.coreAlignmentSupport=true`

```
test.nebula:a:0.15.0
   variant "default" [
      org.gradle.status = release (not requested)
   ]
   Selection reasons:
      - Was requested
      - By constraint : belongs to platform aligned-platform:rules-0:1.0.0
      - Selected by rule

test.nebula:a:1.0.0 -> 0.15.0
\--- compile

test.nebula:b:1.0.0
   variant "default" [
      org.gradle.status = release (not requested)
   ]
   Selection reasons:
      - Was requested
      - By constraint : belongs to platform aligned-platform:rules-0:1.0.0
      - By conflict resolution : between versions 0.15.0 and 1.0.0

test.nebula:b:0.15.0 -> 1.0.0
\--- compile

test.nebula:c:1.0.0
   variant "default" [
      org.gradle.status = release (not requested)
   ]
   Selection reasons:
      - Was requested
      - By constraint : belongs to platform aligned-platform:rules-0:1.0.0

test.nebula:c:1.0.0
\--- compile
```

**Comment by melix** [comment]
Hi @chali ,

So I looked at this issue today and I'm not sure what we should do here. Basically, I think Gradle behaves as expected. The docs for `useVersion` say that this method should precisely override whatever version was selected by the engine, and states that this is what `force()` does.

So what happens in the engine is that whenever we see the version `1.0.0`, it is _substituted with_ `0.15`. This is not a conflict resolution, this is really a substitution in the graph. So the other modules get aligned on `1.0`, but this one does not, specifically because of the `useVersion` call.

Now, if you think of what we implemented last week, things get even more complicated: what we implemented is that if you have a member of a platform which is _forced_ to a version, we _downgrade_ the whole platform. If you think about this and the fact that `useVersion` is supposed to behave a bit like force, then it means we should probably downgrade the whole platform to 0.15 in this case. That is, we should resolve everything to 0.15, not 1.0 like the Nebula plugin. That, however, would also be a breaking change compared to what you have today.

Long story short: I'm not sure what to do, and not sure if we should even do something. It feels to me that the current behavior is semantically correct. What do you think?

**Comment by chali** [comment]
I will research how much this could be impacting our projects. It looks like the other abilities of resolution strategy like substitution are working as we expect. If `useVersion` on one library would bring the whole group down like force that could be reasonable.

**Comment by ljacomet** [comment]
I have been looking into this problem some more. The issue we have is _interpreting intent_ of the rule.
With the rule in the sample:
```
resolutionStrategy.eachDependency { details ->
    if (details.requested.name == 'a') {
        details.useVersion '0.15.0'
    }
}
```
Us developers can extract intent pretty clearly since the test does not contain a version. However inside the Gradle engine we do not have that information.

For example, if I were to change the rule to:
```
resolutionStrategy.eachDependency { details ->
    if (details.requested.name == 'a' && details.requested.version == '0.20.0') {
        details.useVersion '0.15.0'
    }
}
```
In that specific case, ending with aligned modules on `1.0.0` is not an issue and seems to be the right answer.

And because of that inability to extract intent from code, which is not the case when you do a `force`, either through the `resolutionStrategy` or on the dependency itself, I am not sure we can come up with a general solution to this problem for substitution rules.

Interested in hearing feedback on the comments above, both conceptually and technically.

**Comment by chali** [comment]
We had a conversation with @ljacomet and he will investigate what we could do in a case of using substitution API.

**Comment by ljacomet** [comment]
So I confirm that when using the substitution API as follows:
```
resolutionStrategy.dependencySubstitution {
    substitute module('org:a') with module('org:a:0.15.0')        // 1
    substitute module('org:b:0.20.0') with module('org:b:0.15.0') // 2
    substitute module('org:c') with module('org:d:1.0')           // 3
}
```
we can improve the situation for alignment in case 1.
Gradle effectively has all the information it needs to match this substitution with a `force` as the outcome is indeed the same.
Note that neither case 2 nor 3 have that property and thus will not force alignment, instead the substitution will apply.

And obviously, `substitute` that target projects are a different beast.

Also, due to the other things put in place in alignment, having competing `substitute` or `force` or `enforcedPlatform` to different versions of an alignment target will cause the build to fail, indicating that there are multiple forced version for a given platform.

**Comment by ljacomet** [comment]
The support for detecting substitutions as a force and using that information for alignment will be in the 5.0 RC1 release, due really soon.

**Comment by chali** [comment]
Closing this one as there no expected work to be done. Right now I don't have any signal about the strong impact on us so we can reopen discussion when we will have any new information.
